x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlCCSUtils.java

```diff
@@ -190,16 +198,17 @@ static void updateExecutionInfoWithClustersWithNoMatchingIndices(
         IndexResolution indexResolution,
         QueryBuilder filter
     ) {
-        Set<String> clustersWithResolvedIndices = new HashSet<>();
-        // determine missing clusters
+        final Set<String> clustersWithNoMatchingIndices = new HashSet<>(executionInfo.clusterAliases());
         for (String indexName : indexResolution.resolvedIndices()) {
-            clustersWithResolvedIndices.add(RemoteClusterAware.parseClusterAlias(indexName));
+            clustersWithNoMatchingIndices.remove(RemoteClusterAware.parseClusterAlias(indexName));
         }
-        Set<String> clustersRequested = executionInfo.clusterAliases();
-        Set<String> clustersWithNoMatchingIndices = Sets.difference(clustersRequested, clustersWithResolvedIndices);
-        clustersWithNoMatchingIndices.removeAll(indexResolution.unavailableClusters().keySet());
-
-        /**
+        clustersWithNoMatchingIndices.removeAll(EsqlCCSUtils.determineUnavailableRemoteClusters(indexResolution.failures()).keySet());
+        for (var e : indexResolution.failures().entrySet()) {
+            if (e.getValue().stream().anyMatch(f -> f.getException() instanceof IndexNotFoundException == false)) {
+                clustersWithNoMatchingIndices.remove(e.getKey());
```

Comment:
Not sure I understand the logic here - if we didn't find any indices on the cluster for some other reason than `IndexNotFoundException` - wouldn't it still be the same logic? I am not sure I full understand the semantics of `IndexNotFoundException` but if say the index is there but closed or somehow otherwise inaccessible - wouldn't we still want to skip this cluster?

Reply:
see https://github.com/elastic/elasticsearch/pull/130840/commits/3eedd024ec20aace44e2c19f19af75df3a7914fd